### PR TITLE
Add rules for not allowing chained arrows

### PR DESCRIPTION
**Pull request by RickyCook**
```
const a = b => c => `${b}${c}`;
```
```
  1:16  warning  Chained arrow function return needs visual delimiters '(' and ')'  @getify/proper-arrows/return
```

---
This change is [Reviewable](https://reviewable.io/reviews/bluebikesolutions/eslint-config-dutyofcare/1)

### CODE REVIEW

**Comment by RickyCook** [comment]
note: this is better, but still not quite... my objection is with the double function call not being clear

**Comment by daemonl** [comment]
I think it would be worth researching this syntax convention before rejecting it, there are people out there (myself included) who believe this is a really clear approach to a function creating a closure for another function. Being that there is only one body, it's clear that all variables in the chain are in-scope in the inner-most body, so it makes it clear that the outer function call is only there to capture the variable, not to add functionality.

**Comment by RickyCook** [comment]
and i disagree with that, because a function that returns a function is already a non-trivial bit of logic to skim: you should make it as clear as possible what's happening... either that it's returning something that should be passed elsewhere (thereby making the "return" actually a 2-parter), or for some reason you need to call it twice in one place, making it clear that the this needs to happen

it's absolutely _not_ clear what happens here, especially when there's some `async` involved too... you have to really work at it when thinking about async

**Comment by daemonl** [comment]
I think that the issue here is that you aren't really used to it, not that it's intrinsically unclear.

There is no way to agree on code style rules, it's an argument as old as code itself, so I'd really like to minimize 'preferences' as PRs without some way to determine if it's better or worse, beyond us just saying words at each other.

**Comment by RickyCook** [comment]
it's not that im not used to it at all... as ive said, it gets really hard to read through and think about, _especially_ with async mixed in

**Comment by daemonl** [comment]
I'll refer back to 'no way to agree': we need a way to agree.

If this were a PR to add or fix functionality in our application to meet business requirements, I wouldn't block it, as it's preference based, but being that this whole repo is basically for enforcing our preferences on top of AirBnB's preferences and baseline 'not crap code' rules, I don't think just approving it because it's open is the right move.

**Comment by daemonl** [comment]
Can I suggest an alternate solution, I think if we were to remove the rule which enforces the removal of `{}` when the function only has a return should be removed, then in some cases we can 'spell it out' where appropriate.

The problem statement:
```
const contextWrapper => context => specifics => {
  // do things with context and specifics
}
```

Could be represented as:

```
const contectWrapper = contest => {
    return specifics => {
      // do things with context and specifics
  }
}
```

However the current eslint rules will 'auto fix' that by removing the return.

IMO the second is less clear, but at least if we have the option we can use it where appropriate.

A third option to always work around it is to split out into multiple functions, but I don't think that makes it clearer either

```
const contextWrapper = context => specifics => doer(context, specifics)
```

or to not use arrow functions at all

```
function contextWrapper(context) {
  return function(specifics) {
      // do things with context and specifics
  }
}
```

The core issue doesn't seem to be the syntax, but the general concept of wrapping a function in a closure with the context, so I'm not sure if these solve anything which needs to be solved

**Comment by daemonl** [comment]
[ch448]

**Comment by RickyCook** [comment]
the issue i have is with having a single line with all the allows and equals', it makes it very difficult for me to parse as opposed to having them on multiple lines...

How I'd really prefer it:
```
function contextWrapper(context) {
  return specifics => {
    // do things with context and specifics
  }
}
```
```
function contextWrapper(context) {
  return function(specifics) {
    // do things with context and specifics
  }
}
```
```
const contextWrapper = function(context) {
  return specifics => {
    // do things with context and specifics
  }
}
```
Not a great fan of these, because I'm unsure it provides much over just a function, and reduces clarity:
```
const contextWrapper =>
  context =>
    specifics => {
      // do things with context and specifics
    }
```
```
const contextWrapper = context => {
  return specifics => {
    // do things with context and specifics
  }
}
```

**Comment by daemonl** [comment]
I'm not loving the `(` to replace what I think we both are thinking should be a `{`

In the case of the PR which triggered this discussion, it's a builder which builds the interface AWS Lambda is expecting, the return value must be in that format, and the context wrapper is required to configure the body of the callback.

I'm not suggesting that this should be a pattern we use wherever possible, but there is a generic case it is solving, which is giving 'config' to a function using a closure, and I think it's a good pattern to keep available.

In some cases, it will be more appropriate to capture the config in a bound 'this' in a class method, but that would be overkill if the output is only one function and the config / context is small.
```
class doer {
  constructor(config) {
    this.config = config;
    this.actualDoer = this.actualDoer.bind(this);
  }
  actualDoer(requestSpecific) {
    // use this.config and requestSpecific
  }
}
```

The third option is mutable globals, which as you know I strongly oppose, all that's doing is hiding this complexity at the module level.
```
const CONFIG = {}

export function setup(config) {
  CONFIG.requester = config.requester
   // or object.assign, more likely more logic, like libdoc setup
}

function doer(requestSpecific) {
  // do things with CONFIG and requestSpecific
}
```

In summary, I'm saying there are two potential use cases for a pattern like this, one is configuration, the other is meeting an existing interface.

If the pattern itself is problematic, then I don't think we should be looking at the syntax (the rule is probably more 'don't wrap a function in another NOP function), but if we want the pattern and start to use it more broadly, I think having a nice familiar syntax which we can both immediately recognize will be important.

**Comment by RickyCook** [comment]
not doing this
